source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2402,19 +2453,27 @@ void COpenGLDriver::setRenderStates3DMode()
 		ResetRenderStates = true;
 	}
 
+    //actually i'm commenting-out ALMOST all in this if-statemenet
 	if (ResetRenderStates || LastMaterial != Material)
 	{
 		// unset old material
 
-		if (LastMaterial.MaterialType != Material.MaterialType &&
+		/* TODO what OnUnsetMaterial() is responsible for? Can i remove this?
+        if (LastMaterial.Pipeline != Material.Pipeline &&
 				static_cast<uint32_t>(LastMaterial.MaterialType) < MaterialRenderers.size())
-			MaterialRenderers[LastMaterial.MaterialType].Renderer->OnUnsetMaterial();
+			MaterialRenderers[LastMaterial.MaterialType].Renderer->OnUnsetMaterial();*/
 
 		// set new material.
-		if (static_cast<uint32_t>(Material.MaterialType) < MaterialRenderers.size())
-			MaterialRenderers[Material.MaterialType].Renderer->OnSetMaterial(
+		//if (static_cast<uint32_t>(Material.MaterialType) < MaterialRenderers.size())
+			MaterialRenderers[0].Renderer->OnSetMaterial(
 				Material, LastMaterial, ResetRenderStates, this);
 
+            extGlBindProgramPipeline(getPipeline(Material.Pipeline));
+/*
+        WARNING! Commented-out this transform-feedback-related piece of code 
+        because i have no idea about this and i have no time to make it work with E_MATERIAL_TYPE->array_of_shaders changes
+        Plus AFAIK it's going to be removed from the engine anyway..
+
 		if (found->CurrentXFormFeedback&&found->XFormFeedbackRunning)
         {
             if (Material.MaterialType==found->CurrentXFormFeedback->getMaterialType())
```

Comment:
would have  been a check for matching pipelines, basically can't change shader in the middle of xform feedback